File: YuGiOhCard/Controller/PackDetailViewController.m
#import "PackDetailViewController.h"
#import "UrlConsts.h"
#import "FileUtils.h"
#import "PackItem.h"
#import "DatabaseUtils.h"
#import "CardViewController.h"

@interface PackDetailViewController () {
    NSString * _packages;
    NSString * _data_path;
    PackageCards * _pack_cards;
    NSMutableArray * _cards;
}

@end

@implementation PackDetailViewController

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    _packages = [NSString stringWithFormat:@"%@.pkg", self.packageId];
    _data_path = @"data";
    _pack_cards = [[PackageCards alloc] init];
    _cards = [[NSMutableArray alloc] init];
    NSString * jsonData = [FileUtils readTextFile:_packages loadPath:_data_path];
    if ([jsonData isEqualToString:@""]) {
        HttpUtils * hu = [HttpUtils alloc];
        hu.delegate = self;
        NSString * packageCardUrl = [NSString stringWithFormat:URL_PACAKGE_CARD, self.packageId];
        [hu get:packageCardUrl];
    } else {
        [self loadData:jsonData];
    }
    
    [self.refreshButton setTarget:self];
    [self.refreshButton setAction:@selector(refreshClicked:)];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}


#pragma mark - action

-(void) refreshClicked: (id) sender {
    HttpUtils * hu = [HttpUtils alloc];
    hu.delegate = self;
    NSString * packageCardUrl = [NSString stringWithFormat:URL_PACAKGE_CARD, self.packageId];
    [hu get:packageCardUrl];
}

#pragma mark -http

-(void) httpUtils:(HttpUtils *)httpUtils receivedData:(NSData *)data {
    if (data != nil) {
        NSString * json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        [FileUtils writeTextFile:_packages savePath:_data_path fileContent:json];
        [self loadData:json];
    }
}

-(void) httpUtils:(HttpUtils *)httpUtils receivedError:(NSString *)err {
    
}

-(void) loadData: (NSString *) json {
    NSData * data = [json dataUsingEncoding:NSUTF8StringEncoding];
    id pack = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:nil];
    _pack_cards.name = pack[@"name"];
    [_pack_cards.cards removeAllObjects];
    [_pack_cards.cards addObjectsFromArray:pack[@"cards"]];
    _cards = [DatabaseUtils queryCardsViaIds:_pack_cards.cards];
    
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return _cards.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    CardItem * item = _cards[indexPath.row];
    cell.backgroundColor = [UIColor clearColor];
    cell.textLabel.textColor = [UIColor whiteColor];
    cell.textLabel.text = item.name;
    cell.detailTextLabel.text = item.sCardType;
    
    return cell;
}


#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"pushCard"]) {
        CardItem * item = _cards[[self.tableView indexPathForSelectedRow].row];
        [[segue destinationViewController] setCardId:item.card_id];
        [[segue destinationViewController] setCardName:item.name];
    }
}


@end

File: swift/YuGiOhCard/Controller/PackDetailViewController.swift
import UIKit

class PackDetailViewController: UITableViewController, HttpUtilsDelegate {

    @IBOutlet var refreshButton: UIBarButtonItem?
    var packageId: String?
    var packageName: String?
    
    var _packages: String?
    var _data_path: String?
    var _pack_cards: PackageCards?
    var _cards: NSMutableArray?
    
    override func viewWillAppear(animated: Bool) {
        UIUtils.setStatusBar(true)
        UIUtils.setNavBar(self.navigationController!.navigationBar)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = self.packageName!
        _packages = "\(self.packageId).pkg"
        _data_path = "data"
        _pack_cards = PackageCards()
        _cards = NSMutableArray()
        let jsonData = FileUtils.readTextFile(_packages!, loadPath:_data_path!)
        if (jsonData == "") {
            let hu = HttpUtils()
            hu.delegate = self
            let packageCardUrl = NSString(format: URL_PACAKGE_CARD, self.packageId!) as String
            hu.get(packageCardUrl)
        } else {
            self.loadData(jsonData)
        }
        
        self.refreshButton!.target = self
        self.refreshButton!.action = "refreshClicked:"
        self.tableView.tableFooterView = UIView(frame: CGRectZero)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func refreshClicked(sender: AnyObject) {
        let hu = HttpUtils()
        hu.delegate = self
        let packageCardUrl = NSString(format: URL_PACAKGE_CARD, self.packageId!) as String
        hu.get(packageCardUrl)
    }
    
    func httpUtils(httpUtils: HttpUtils, receivedData data: NSData?) {
        if (data != nil) {
            let json = NSString(data: data!, encoding:NSUTF8StringEncoding)
            FileUtils.writeTextFile(_packages!, savePath:_data_path!, fileContent:json as! String)
            self.loadData(json as! String)
        }
    }
    
    func httpUtils(httpUtils: HttpUtils, receivedError err: NSString) {
        
    }
    
    func loadData(json: String) {
        let data = (json as NSString).dataUsingEncoding(NSUTF8StringEncoding)
        let pack: AnyObject? = try? NSJSONSerialization.JSONObjectWithData(data!, options:NSJSONReadingOptions.MutableLeaves)
        if (pack != nil) {
            _pack_cards!.name = (pack! as! NSDictionary).objectForKey("name") as! String
            _pack_cards!.cards.removeAllObjects()
            _pack_cards!.cards.addObjectsFromArray((pack! as! NSDictionary).objectForKey("cards") as! [AnyObject])
            _cards = DatabaseUtils.queryCardsViaIds(_pack_cards!.cards)
            self.tableView.reloadData()
        }
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _cards!.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath:indexPath) 
        let item = _cards![indexPath.row] as! CardItem
        cell.backgroundColor = UIColor.clearColor()
        cell.textLabel!.textColor = UIColor.whiteColor()
        cell.textLabel!.text = item.name
        cell.detailTextLabel!.text = item.sCardType
        return cell
    }
    
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.tableView.deselectRowAtIndexPath(indexPath, animated:false)
        let item = _cards![indexPath.row] as! CardItem
        PushUtils.pushCard(item, navController:self.navigationController!)
    }
    
}
